tensorflow/lite/experimental/shlo/ops/dot_general.cc

```diff
@@ -0,0 +1,638 @@
+            rhs.quantized_per_axis_element_type().ExpressedType()) {
+          if (lhs.quantized_per_tensor_element_type().ExpressedType() !=
+              output.quantized_per_axis_element_type().ExpressedType()) {
+            return absl::FailedPreconditionError(
+                "stablehlo.dot_general: The expressed_type of output must be "
+                "the same as the expressed_type of lhs and rhs.");
+          }
+        }
+      }
+      auto check_zero_points_values_are_zero =
+          [](const auto& zero_points) -> bool {
+        return std::visit(
+            [](const auto& v) {
+              return std::all_of(v.begin(), v.end(), [](const auto value) {
+                return value == static_cast<decltype(value)>(0);
+              });
+            },
+            zero_points);
+      };
+      if (!check_zero_points_values_are_zero(
+              rhs.quantized_per_axis_element_type().ZeroPoints())) {
+        return absl::FailedPreconditionError(
+            "stablehlo.dot_general: The rhs per-axis should have zero points "
+            "as 0.");
+      }
+    } else if (rhs.IsPerAxisQuantized()) {
+      for (DimensionSize i = 0; i < rhsc_size; ++i) {
+        if (rhs_contracting_dimensions[i] ==
+            rhs.quantized_per_axis_element_type().QuantizedDimension()) {
+          return absl::FailedPreconditionError(
+              "stablehlo.dot_general: If the rhs is per-axis quantized than "
+              "the quantization_dimensions of rhs should not be in "
+              "rhs_contracting_dimensions.");
+        }
+      }
+    }
+  }
+  return absl::OkStatus();
+}
+
+template <DataType storage_type>
+absl::Status EvaluateImpl(DotGeneralOp& op, const Tensor& lhs,
+                          const Tensor& rhs,
+                          absl::Span<int64_t> lhs_batching_dimensions,
+                          absl::Span<int64_t> rhs_batching_dimensions,
+                          absl::Span<int64_t> lhs_contracting_dimensions,
+                          absl::Span<int64_t> rhs_contracting_dimensions,
+                          Tensor& output) {
+  using StorageT = StorageType<storage_type>;
+  const StorageT* lhs_data = lhs.GetDataAs<storage_type>();
+  const StorageT* rhs_data = rhs.GetDataAs<storage_type>();
+  StorageT* output_data = output.GetDataAs<storage_type>();
+  const DimensionSize lhs_size = lhs.NumElements();
+  const DimensionSize rhs_size = rhs.NumElements();
+  const DimensionSize output_size = output.NumElements();
+  const size_t lhs_rank = lhs.Rank();
+  const size_t rhs_rank = rhs.Rank();
+  const size_t output_rank = output.Rank();
+  const DimensionSize lhsb_size = lhs_batching_dimensions.size();
+  const DimensionSize lhsc_size = lhs_contracting_dimensions.size();
+
+  // function to generate indices for output
+  auto GenerateIndices = [&](size_t index) -> void {
+    size_t rank = op.output_shape.size();
+    size_t divisor = 1;
+    for (size_t i = 0, j = rank - 1; i < rank; ++i, --j) {
```

Comment:
This for loop doesn't use `i`, it would be more understandable as `for (size_t j = rank - 1; j >= 0; --j)`